src/Microsoft.AspNetCore.Mvc.Core/Internal/LocalRedirectResultExecutor.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    public class LocalRedirectResultExecutor
+    {
+        private readonly ILogger _logger;
+        private readonly IUrlHelperFactory _urlHelperFactory;
+
+        public LocalRedirectResultExecutor(ILoggerFactory loggerFactory, IUrlHelperFactory urlHelperFactory)
+        {
+            if (loggerFactory == null)
+            {
+                throw new ArgumentNullException(nameof(loggerFactory));
+            }
+
+            if (urlHelperFactory == null)
+            {
+                throw new ArgumentNullException(nameof(urlHelperFactory));
+            }
+
+            _logger = loggerFactory.CreateLogger<LocalRedirectResultExecutor>();
+            _urlHelperFactory = urlHelperFactory;
+        }
+
+        public void Execute(ActionContext context, LocalRedirectResult result)
+        {
+            var urlHelper = result.UrlHelper ?? _urlHelperFactory.GetUrlHelper(context);
+
+            // IsLocalUrl is called to handle  Urls starting with '~/'.
```

Comment:
Could we just change this to say "Only allow local urls"? What is considered a `IsLocalUrl` depends on the implementation of `IUrlHelper`.